client/clientutil/modelinfo.go

```diff
@@ -0,0 +1,580 @@
+	} else {
+		mainAssertion = modelAssertion
+	}
+
+	serializeJSON := func(v interface{}) error {
+		marshalled, err := json.MarshalIndent(v, "", "  ")
+		if err != nil {
+			return err
+		}
+		_, err = w.Write(marshalled)
+		w.Flush()
+		return err
+	}
+
+	if assertion {
+		// if we are using the serial assertion and we specifically didn't find the
+		// serial assertion, bail with specific error
+		if useSerial && serialAssertion == nil {
+			return errNoMainAssertion
+		}
+
+		modelJSON := modelAssertJSON{}
+		modelJSON.Headers = mainAssertion.Headers()
+		modelJSON.Body = string(mainAssertion.Body())
+		return serializeJSON(modelJSON)
+	}
+
+	modelData := make(map[string]interface{})
+	if useSerial && serialAssertion == nil {
+		// for serial assertion, the primary keys are output (model and
+		// brand-id), but if we didn't find the serial assertion then we still
+		// output the brand-id and model from the model assertion, but also
+		// return a devNotReady error
+		modelData["brand-id"] = modelAssertion.HeaderString("brand-id")
+		modelData["model"] = modelAssertion.HeaderString("model")
+		if err := serializeJSON(modelData); err != nil {
+			return err
+		}
+		return errNoSerial
+	}
+
+	// ordering of the primary keys for model: brand, model, serial
+	// ordering of primary keys for serial is brand-id, model, serial
+
+	// output brand/brand-id
+	modelData["brand-id"] = mainAssertion.HeaderString("brand-id")
+	modelData["model"] = mainAssertion.HeaderString("model")
+
+	// for the serial header, if there's no serial yet, it's not an error for
+	// model (and we already handled the serial error above) but need to add a
+	// parenthetical about the device not being registered yet
+	var serial string
+	if serialAssertion == nil {
+		// For JSON verbose is always true, so we need to escape the dash
+		serial = modelFormatter.GetEscapedDash()
+		serial += " (device not registered yet)"
+	} else {
+		serial = serialAssertion.HeaderString("serial")
+	}
+
+	// only output the grade if it is non-empty, either it is not in the model
+	// assertion for all non-uc20 model assertions, or it is non-empty and
+	// required for uc20 model assertions
+	grade := modelAssertion.HeaderString("grade")
+	if grade != "" {
+		modelData["grade"] = grade
+	}
+
+	storageSafety := modelAssertion.HeaderString("storage-safety")
+	if storageSafety != "" {
+		modelData["storage-safety"] = storageSafety
+	}
+
+	// serial is same for all variants
+	modelData["serial"] = serial
+	allHeadersMap := mainAssertion.Headers()
+
+	// always print extra information for JSON
+	for _, headerName := range niceOrdering {
+		invalidTypeErr := fmt.Errorf(invalidTypeMessage, headerName)
+
+		headerValue, ok := allHeadersMap[headerName]
+		// make sure the header is in the map
+		if !ok {
+			continue
+		}
+
+		// switch on which header it is to handle some special cases
+		switch headerName {
+		// list of scalars
+		case "required-snaps", "system-user-authority":
```

Comment:
I think most of this code can be simplified for the JSON case, apart device-key that we need to think a bit about the rest of the code can simply copy the value across into modelData afaict

Reply:
Ah nice, good to know, that will simplify the JSON one a lot

Reply:
Just to be clear:

"required-snaps", "system-user-authority" 
I have removed those two

"timestamp"
This still gets reformatted, should I remove this?

"snaps"
This filters out members, should we keep or remove this?

Reply:
> "timestamp"
> This still gets reformatted, should I remove this?

we should let JSON it's formatting I think

> "snaps"
> This filters out members, should we keep or remove this?

what members is it actually filtering it out? (I think it's doing members one by one to do formatting, but maybe I'm misremembering)